JSUI-2798 Improved accessibility of `Quickview`

btaillon opened the pull request:
https://coveord.atlassian.net/browse/JSUI-2798

1. Added label to close button
2. Added tab index to close button for keyboard navigation
3. Locked keyboard focus within `QuickView`'s modal while it's open.
4. Added accessibility tests for QuickView

This was tested with Android's TalkBack and doesn't seem to work. Most tutorials and articles online, even when searching specifically for accessible modals, seem to ignore accessibility altogether by detecting the `tab` key. However, some tutorials using ally.js seem to detect focus/blur for the desktop and add `aria-hidden` for mobile accessibility tools, which seems to be a less radical solution than setting `tabindex="-1"` to every focusable element. There is no guarantee that this works with iOS's VoiceOver either, but setting `aria-hidden` and `aria-modal` seems to be the recommended method by W3 (https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html).

[![Deploy](https://www.herokucdn.com/deploy/button.svg)](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

btaillon commented on lib/modal-box/index.d.ts:
```diff
@@ -1,96 +1,97 @@
-declare module Coveo.ModalBox {
+declare namespace Coveo.ModalBox {
```
I tried to revert this change along with the tabbing, but it looks like whenever I commit my changes, the file gets re-formatted like this. I'll write comments under the lines I changed in this file.

btaillon commented on lib/modal-box/index.d.ts:
```diff
@@ -1,96 +1,97 @@
+    /**
+     * Specify the title of the modal box
+     */
+    title?: string | HTMLElement;
```
This is one of the changes I made. I changed this type from `string` to `string | HTMLElement` since some old code was already setting an `HTMLElement` on this property.

btaillon commented on lib/modal-box/index.d.ts:
```diff
@@ -1,96 +1,97 @@
+     * Specify the content that you wish to put inside the modal box
+     */
+    body?: HTMLElement;
+    sizeMod?: 'big';
```
This is the other change I made: I added this missing property to avoid getting errors. I don't know what other values it can get than `big`, so I only set it to the type `'big'` for now.

btaillon commented on src/ui/Quickview/Quickview.ts:
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
+      }
+      if (e.keyCode !== KEYBOARD.TAB) {
+        return;
+      }
```
I'm worried this might not work on some screen readers. Is there some tools I should test this with?

olamothe replied:
You can use osx voiceover.

olamothe commented on accessibilityTest/AccessibilityQuickview.ts:
```diff
@@ -1,33 +1,101 @@
+  }
+  return result;
+}
+
```
Could simply use `Dom.nodeListToArray` instead of rewriting it ?

btaillon replied:
Woops, I wasn't aware this function existed. I'll use it instead.

olamothe commented on accessibilityTest/AccessibilityQuickview.ts:
```diff
@@ -1,33 +1,101 @@
+
+    it('should have a close button when opened', async done => {
+      await openQuickview();
+      expect(document.querySelector('.coveo-small-close')).not.toBeNull();
```
I would create a function for `document.querySelector('.coveo-small-close')` -> `getQuickviewCloseButton()`

btaillon replied:
Good idea, fixing that.

olamothe commented on accessibilityTest/AccessibilityQuickview.ts:
```diff
@@ -1,33 +1,101 @@
+          break;
+        }
+      }
+      expect(document.activeElement.outerHTML).toBe(initialFocus.outerHTML);
```
Can you instead test directly for `expect(document.activeElement).toBe(initialFocus)` ? Instead of outerHTML ?

btaillon replied:
Woops, I used this because jasmine gave ambiguous error messages otherwise. I forgot to fix this, I'll remove `.outerHTML`.

olamothe commented on src/ui/Quickview/Quickview.ts:
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
+    return _.sortBy(container.querySelectorAll<HTMLElement>('[tabindex]'), element => element.tabIndex);
+  }
+
+  private createFocusTrap(element: HTMLElement) {
```
Was creating such a focusTrap the recommended approach on dequeue university ?

Also, this whole function is pretty hard to follow/read.

Id split/break the logic up in different method. 
I could even see all of this moved to `FocusTrap` module/class maybe ?

Which accept a `container` or root element inside which the trap should be created, and the tab cycling logic would be implemented.

btaillon replied:
I couldn't find any recommendation from Deque other than to make sure elements outside of a modal aren't focusable.

> If a screen reader user navigates by elements such as headings, landmarks, links, form elements, etc., the only elements available when the dialog is open are the elements in the dialog itself. There must not be a way to access anything outside of the dialog with screen reader keyboard shortcuts.

As for the code being hard to read, I agree that it could be broken up into a separate class and I'll do that.

olamothe commented on src/ui/Quickview/Quickview.ts:
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
+      if (!element.contains(e.target as HTMLElement)) {
+        const focusableElements = this.getFocusableElements(element);
+        focusableElements[0].focus();
+        e.preventDefault();
```
Why is the `preventDefault` needed here ?

btaillon replied:
`preventDefault` is needed here to prevent the browser from changing the currently focused element to one that's outside of the `Quickview`.

olamothe commented on src/ui/Quickview/Quickview.ts:
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
+        return;
+      }
+      if (!element.contains(e.target as HTMLElement)) {
+        const focusableElements = this.getFocusableElements(element);
```
What if there's no focusable elements returned ?

btaillon replied:
Since the focus trap is currently only implemented for the `Quickview`, it always has at-least one focusable element (the close button). However, I agree that it would be a good idea to add a check here. I'll add one.

olamothe commented on src/ui/Quickview/Quickview.ts:
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
+      } else {
+        if (activeElement === focusableElements[0]) {
+          _.last(focusableElements).focus();
+          e.preventDefault();
```
Need to make sure it won't break if there's somehow no focusable element.

btaillon replied:
I'll fix that.

olamothe commented on src/ui/Quickview/Quickview.ts:
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
+  }
+
+  private createFocusTrap(element: HTMLElement) {
+    document.addEventListener('focusin', e => {
```
This function will get created everytime a new quickview is opened, but it never gets cleaned up when the modal box closes. That's problematic.

btaillon replied:
Woops, I didn't think about the fact that those are now on `document` and need to be cleaned up. I'm going to fix this.

olamothe commented on src/ui/Quickview/Quickview.ts:
```diff
@@ -453,10 +464,64 @@ export class Quickview extends Component {
+        return;
+      }
+    });
+    $$(document.body).on('keydown', (e: KeyboardEvent) => {
```
Same thing about the function being created everytime a new quickview is created but never cleaned up.

btaillon replied:
I'll clean up function too.

olamothe commented on unitTests/Simulate.ts:
```diff
@@ -221,6 +221,31 @@ export class Simulate {
     return modalBox;
   }
 
+  static fullModalBoxModule(): ModalBox {
```
Why did you create a new function for this instead of modifying the existing `modalBoxModule` ? It seems very hackish ;)

btaillon replied:
I did that because the context I needed to test `Quickview` was too different from the already existing one and I didn't want to risk breaking other accessibility tests. I'm going to test this one with every accessibility test and see if it works.

btaillon commented:
I added the `FocusTrap` component, which should be easier to test and re-use in other components if needed. I also added a few unit tests to go along with it and avoid regressions.

ThibodeauJF commented on src/ui/FocusTrap/FocusTrap.ts:
```diff
@@ -0,0 +1,112 @@
+import _ = require('underscore');
```
Import only the few methods you need for underscore

btaillon replied:
Woops, fixing that.

ThibodeauJF commented on src/ui/FocusTrap/FocusTrap.ts:
```diff
@@ -0,0 +1,112 @@
+
+  private focusFirstElement() {
+    const elements = this.focusableElements;
+    if (elements.length > 0) {
```
could only use `if (elements.length)`

btaillon replied:
Yeah that's what I initially wrote but I wasn't sure which option was preferable. I'll change it to `elements.length`.
